### PR TITLE
[WIP] Reuse GRPC Tunnel for multiple client dial requests

### PR DESCRIPTION
**Pull request by Jefftree**
Reuse a single GRPC Tunnel for multiple client dial requests.

A couple of notes:

- Connection health checking will be required (since we expect one tunnel to persist throughout the duration of the apiserver being active). That is not included in this PR
- `concurrent_test.go` was renamed to `grpc_tunnel_test.go` since it only tested grpc tunnel concurrency.

### CODE REVIEW

File: cmd/client/main.go [hunk 1]
```diff
@@ -264,7 +267,13 @@ func (c *Client) getUDSDialer(o *GrpcProxyClientOptions) (func(ctx context.Conte
...
 		})
-		tunnel, err := client.CreateSingleUseGrpcTunnel(o.proxyUdsName, dialOption, grpc.WithInsecure(), grpc.WithUserAgent(o.userAgent))
+		tunnel, err := client.CreateReusableGrpcTunnel(o.proxyUdsName, dialOption, grpc.WithInsecure(), grpc.WithUserAgent(o.userAgent))
+
+		go func() {
+			<-stopCh
+			tunnel.Close()
+		}()
+
 		if err != nil {
```
Comment (caesarxuchao) [inline cmd/client/main.go:277]: Close the tunnel if err!=nil.

File: konnectivity-client/pkg/client/client.go [hunk 2]
```diff
@@ -49,13 +52,30 @@ type grpcTunnel struct {
 	conns           map[int64]*conn
 	pendingDialLock sync.RWMutex
 	connsLock       sync.RWMutex
+	singleUse       bool
```
Comment (caesarxuchao) [inline konnectivity-client/pkg/client/client.go:55]: Let's remove the singleUse support if Kubernetes moves to use the reusable one.

File: konnectivity-client/pkg/client/client.go [hunk 3]
```diff
@@ -49,13 +52,30 @@ type grpcTunnel struct {
...
+	proxyConn       *grpc.ClientConn
```
Comment (caesarxuchao) [inline konnectivity-client/pkg/client/client.go:56]: Is `grpcConn` clearer?

File: tests/grpc_tunnel_test.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+	"google.golang.org/grpc"
+	"sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client"
+)
+
+// The default IdleConnTimeout is 90s and will cause all Dials
+// using the same Dialer to reuse the initial connection instead
+// of calling the Dial function again. Setting it to 0 forces a new
+// Dial on each new http client
+
```
Comment (caesarxuchao) [inline tests/grpc_tunnel_test.go:19]: Good catch. Please remove the empty line.

**Comment by Jefftree** [comment]
Closing in favor of splitting up this PR.

**Pull request closed without merging**